content/en/blog/low-hanging-performance/index.md

```diff
@@ -0,0 +1,460 @@
+     b.Grow(2 * len(s))
+
+     for i, v := range s {
+          skip := v == '(' || v == ')' || v == '*'
+          if skip {
+               if !prevSkipped && i != 0 { // first time write space
+                    b.WriteRune(' ')
+               }
+               prevSkipped = skip
+               continue
+          }
+          toSpace := v == '.' || v == '_'
+          if toSpace {
+               if prevSkipped {
+                    continue
+               } else if v == '.' {
+                    b.WriteRune(':')
+               }
+               v = ' '
+               prevSkipped = true
+          } else {
+               isUpper = unicode.IsUpper(v)
+               if isUpper {
+                    v = unicode.ToLower(v)
+                    if !prevSkipped && splittable {
+                         b.WriteRune(' ')
+                         prevSkipped = true
+                    }
+               } else {
+                    prevSkipped = false
+               }
+          }
+          b.WriteRune(v)
+          splittable = !isUpper || unicode.IsNumber(v)
+     }
+     return b.String()
+}
+```
+
+Let’s take another example where results are even more drastically faster, but
+the reason is precisely the same. The input set is much smaller than what the
+first implementation function is meant to be used.
+
+The results:
+
+```
+goos: linux
+goarch: arm64
+pkg: github.com/lainio/err2/assert
+BenchmarkGoid_MyByteToInt-5     529940133                2.273 ns/op           0 B/op          0 allocs/op
+BenchmarkGoid_Old-5              3919933               307.0 ns/op            64 B/op          3 allocs/op
+PASS
+ok      github.com/lainio/err2/assert   4.136s
+```
+
+The first implementation:
+
+```go
+func oldGoid(buf []byte) (id int) {
+     _, err := fmt.Fscanf(bytes.NewReader(buf), "goroutine %d", &id)
+     if err != nil {
+          panic("cannot get goroutine id: " + err.Error())
+     }
+     return id
+}
+```
+The above code is quite self explanatory, and that's very good.
+
+The second and ~~faster~~*fastest* implementation:
+
+```go
+func asciiWordToInt(b []byte) int {
+     n := 0
+     for _, ch := range b {
+          if ch == ' ' {
+               break
+          }
+          ch -= '0'
+          if ch > 9 {
+               panic("cannot get goroutine id")
+          }
+          n = n*10 + int(ch)
+     }
+     return n
+}
+```
+
+These two functions do precisely the same thing, or should I say almost because
+the latter’s API is more generic. The converted integer must be the first byte
+in the slice.
+
+It is much over 100x faster! Why? Because the only thing we need is to process
+the ASCII string that comes in byte slice type.
+
+> You might ask whether this ruined the *readability*, which is fair. But no,
+> because the function `asciiWordToInt` is called from `GoroutineID`, which is
+> just enough -- trust abstraction layering. (See the rule #1.)
+
+Next time you are writing something, think twice -- I do ;-)
+
```

Comment:
I think a summary chapter would be a good addition. Just bullets that what did the reader just learned and how to utilize that info in their daily work.